lib/render/mayaToHydra/plugin.cpp

```diff
@@ -64,7 +56,17 @@ PLUGIN_EXPORT MStatus initializePlugin(MObject obj) {
         return ret;
     }
 
-    MtohInitializeRenderGlobals();
+    auto plugins = MtohInitializeRenderPlugins();
+    if (auto* renderer = MHWRender::MRenderer::theRenderer()) {
+        for (const auto& desc : plugins.first) {
+            std::unique_ptr<MtohRenderOverride> mtohRenderer(new MtohRenderOverride(desc));
+            renderer->registerOverride(mtohRenderer.get());
+            mtohRenderer.release();
+        }
```

Comment:
@elrond79 Good catch!

It is bad that the API doc for **registerOverride**/**deregisterOverride** doesn't say anything about the lifetime. By looking into the code, the API implementation is clearly assuming that the MRenderOverride instance should still be alive until deregistration is called, so for your first question it is indeed unsafe.

As to the two patterns, the second pattern clearly results in a memory leak. Other than that, I don't have strong preference here. The second pattern avoids caching the pointer, but results in an additional lookup by **findRenderOverride** before deregistration. I think it is cleaner, and the lookup itself should be fine as we don't have a big list of loaded render overrides for most normal use cases (rather than extreme testing).

I will log issues you found here (API doc + mem leak) to Maya team. Thanks.

Reply:
I'm a bit confused, `std::unique_ptr<>::release` does not delete the pointer it holds, it relinquishes ownership.
**registerOverride** is not noexcept, so if it throws, the pointer would be deleted, on the assumption **registerOverride** could not complete the request ?

Reply:
Also there is one cache, in __rendererAttributes_ as it is now. The list that is built up during `MtohInitializeRenderPlugins` is transferred to **MtohInitializeRenderGlobals** via the std::move().

I can see if this can be avoided by doing what `MtohInitializeRenderGlobals` does in `MtohInitializeRenderPlugins` , but have a vague feeling there's trouble involved in moving __rendererAttributes_ to a different module (hence the lambdas from the first approach).

Reply:
Aah, I thought `std::unique_ptr<>::release` would delete the object thus raising the concern. It sounds good to me then.
The only suggestion is to add a comment here to explain your thinking, esp. about the extra protection in case there is exception.

Reply:
The Maya side issue (missing API doc about lifetime and devkit memory leak) has been logged with MAYA-103639.

Reply:
Doh - my mistake (and it was my initial mistake that probably misled Huidong)! Anyway, yeah, doing this for exception cleanup makes sense - let's just add the comment, then!

Reply:
Also - @HdC-adsk  - thanks for making the maya ticket, and I'll make a separate PR to properly clean up the MRenderOverride in uninitialize.